src/ai/ai_gui.cpp

```diff
@@ -264,6 +267,354 @@ struct AIConfigWindow : public Window {
+	}
+
+	void OnClick(Point pt, int widget, int click_count) override
+	{
+		switch (widget) {
+			case WID_AIS_BACKGROUND: {
+				Rect r = this->GetWidget<NWidgetBase>(widget)->GetCurrentRect().Shrink(WidgetDimensions::scaled.matrix, RectPadding::zero);
+				int num = (pt.y - r.top) / this->line_height + this->vscroll->GetPosition();
+				if (num >= (int)this->visible_settings.size()) break;
+
+				VisibleSettingsList::const_iterator it = this->visible_settings.begin();
+				for (int i = 0; i < num; i++) it++;
+				const ScriptConfigItem config_item = **it;
+				if (!this->IsEditableItem(config_item)) return;
+
+				if (this->clicked_row != num) {
+					this->CloseChildWindows(WC_QUERY_STRING);
+					HideDropDownMenu(this);
+					this->clicked_row = num;
+					this->clicked_dropdown = false;
+				}
+
+				bool bool_item = (config_item.flags & SCRIPTCONFIG_BOOLEAN) != 0;
+
+				int x = pt.x - r.left;
+				if (_current_text_dir == TD_RTL) x = r.Width() - 1 - x;
+
+				/* One of the arrows is clicked (or green/red rect in case of bool value) */
+				int old_val = this->script_config->GetSetting(config_item.name);
+				if (!bool_item && IsInsideMM(x, 0, SETTING_BUTTON_WIDTH) && config_item.complete_labels) {
+					if (this->clicked_dropdown) {
+						/* unclick the dropdown */
+						HideDropDownMenu(this);
+						this->clicked_dropdown = false;
+						this->closing_dropdown = false;
+					} else {
+						int rel_y = (pt.y - r.top) % this->line_height;
+
+						Rect wi_rect;
+						wi_rect.left = pt.x - (_current_text_dir == TD_RTL ? SETTING_BUTTON_WIDTH - 1 - x : x);
+						wi_rect.right = wi_rect.left + SETTING_BUTTON_WIDTH - 1;
+						wi_rect.top = pt.y - rel_y + (this->line_height - SETTING_BUTTON_HEIGHT) / 2;
+						wi_rect.bottom = wi_rect.top + SETTING_BUTTON_HEIGHT - 1;
+
+						/* If the mouse is still held but dragged outside of the dropdown list, keep the dropdown open */
+						if (pt.y >= wi_rect.top && pt.y <= wi_rect.bottom) {
+							this->clicked_dropdown = true;
+							this->closing_dropdown = false;
+
+							DropDownList list;
+							for (int i = config_item.min_value; i <= config_item.max_value; i++) {
+								list.emplace_back(new DropDownListCharStringItem(config_item.labels->Find(i)->second, i, false));
+							}
+
+							ShowDropDownListAt(this, std::move(list), old_val, -1, wi_rect, COLOUR_ORANGE, true);
+						}
+					}
+				} else if (IsInsideMM(x, 0, SETTING_BUTTON_WIDTH)) {
+					int new_val = old_val;
+					if (bool_item) {
+						new_val = !new_val;
+					} else if (x >= SETTING_BUTTON_WIDTH / 2) {
+						/* Increase button clicked */
+						new_val += config_item.step_size;
+						if (new_val > config_item.max_value) new_val = config_item.max_value;
+						this->clicked_increase = true;
+					} else {
+						/* Decrease button clicked */
+						new_val -= config_item.step_size;
+						if (new_val < config_item.min_value) new_val = config_item.min_value;
+						this->clicked_increase = false;
+					}
```

Comment:
Possible programmatic simplification (although arguably makes it less human-readable):
```suggestion
					} else {
						/* Increase/Decrease button clicked */
						this->clicked_increase = (x >= SETTING_BUTTON_WIDTH / 2);
						new_val = this->clicked_increase ?
							std::min(config_item.max_value, new_val + config_item.step_size) :
							std::max(config_item.min_value, new_val - config_item.step_size);
					}
```

Reply:
Same could apply to GSConfigWindow